samples/api/timestamp_queries/README.md

```diff
@@ -0,0 +1,184 @@
+
+Before we can start writing data to the query pool, we need to reset it. When using Vulkan 1.0 or 1.1, this requires us to enable the `VK_EXT_host_query_reset` extension:
+
+```cpp
+add_device_extension(VK_EXT_HOST_QUERY_RESET_EXTENSION_NAME);
+```
+
+With using Vulkan 1.2 this extension has become part of the core and we won't have to manually enable it.
+
+Independent of this, we also need to enable the `hostQueryReset` physical device feature:
+
+```cpp
+auto &requested_extension_features= gpu.request_extension_features<VkPhysicalDeviceHostQueryResetFeaturesEXT>(VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_HOST_QUERY_RESET_FEATURES_EXT);
+requested_extension_features.hostQueryReset = VK_TRUE;
+```
+
+With features and extensions properly enabled, we can now reset the pool at the start of the command buffer, before writing the first timestamp. This is done using `vkCmdResetQueryPool`:
+
+```cpp
+...
+vkBeginCommandBuffer(draw_cmd_buffers[i], &command_buffer_begin_info);
+vkCmdResetQueryPool(draw_cmd_buffers[i], query_pool_timestamps, 0, static_cast<uint32_t>(time_stamps.size()));
+```
+------
+
+## Writing time stamps
+
+Unlike getting CPU side timing information that can be queried immediately, with GPU time stamps we need to tell the implementation inside a command buffer when/where to write timestamps instead. The results are then fetched afterwards (see below).
+
+This is done inside the command buffer with `vkCmdWriteTimestamp`. This function will request a timestamp to be written from the GPU for a certain pipeline stage and write that value to memory.
+
+The most interesting part of calling this function is the `pipelineStage` argument. As noted earlier, it's theoretically possible to use any pipeline stage in here, not all pipeline stages will net proper results due to how GPUs overlap work. It's also important to note that not all implementations are able to latch timers at all pipeline stages (e.g. if they don't have hardware that maps to a given stage) and may return timers at a later pipeline stage instead.
+ 
+Calling this function also defines an execution dependency similar to a barrier on all commands that were submitted before it.
+
+```cpp
+vkCmdWriteTimestamp(draw_cmd_buffers[i], VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, query_pool_timestamps, 0);
+// Do some work
+for (int i = 0; i < draw_call_count; i++) {
+	vkCmdDraw(...);
+}
+vkCmdWriteTimestamp(draw_cmd_buffers[i], VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, query_pool_timestamps, 1);
+```
+
+To measure GPU times for the draw calls(s) we first tell the GPU to write a timestamp at the `VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT` pipeline stage. This is not a real pipeline stage (as in e.g. the vertex or fragment stages) but a special constant that tells the GPU to wait for all commands currently in flight to finish. This ensures, that we get a timestamp right before starting on the draw calls we want to measure. This timestamp will then be the base for calculating a delta time.
```

Comment:
> [TOP is] a special constant that tells the GPU to wait for all commands currently in flight to finish

This seems wrong -- TOP here means wait for all previous commands to finish the TOP stage, which in turns mean... well, nothing, I think, since the TOP stage has no observable effects?

If you want to make sure all previous commands are _done_ executing, shouldn't this be BOTTOM?

Reply:
I think, the usage of TOP and BOTTOM here is correct.
But the wording might be misleading: the GPU doesn't wait for anything. Writing a timestamp does not imply some waitIdle, or so.
Instead, with TOP, the timestamp is written as soon as all previous commands were taken by the GPU's command processor (not necessarily started yet, though!), and with BOTTOM, the timestamp is written as soon as all previous commands were done.

Reply:
But for the purpose of timing individual workloads, don't you want the two time-stamps to reflect (a) when the previous work ended, and (b) when the work you want to time ended?

A timestamp in TOP can legally be taken immediately after the vkQueueSubmit -- in the same way that e.g. an image transitions with srcStage = TOP can legally be executed immediately after vkQueueSubmit. Furthermore, _all_ timestamps with srcStage = TOP can be taken _at the same time_ immediately after vkQueueSubmit, which does not seem like the intent here.

(That said, it's true that on many GPUs the command processor will not re-order work in this way, so in practice this is not a problem. But there are also implementations _without_ a global command processor, which do re-order.)

That's why it seems like src=BOTTOM followed by src=BOTTOM makes the most sense here, so that you're sure to get:
- Timestamp before the work starts (after previous work is done), and
- Timestamp after the work is done

Reply:
I think, it kind of depends on what you want to measure here... Because of the parallel work scheduling of modern GPUs,
* with TOP to BOTTOM, you might have some work from previous commands on your bill, but can be sure to measure all your work.
* with BOTTOM to BOTTOM, you can be sure to not have any work from previous commands on your bill, but might have missed some of your work.

That is, with modern GPUs, it's hard to have a "Timestamp before the work starts (after previous work is done)", because there might be no point in time that fits that criterium. You actually described exactly that situation in your parenthesized paragraph.
The only way to be sure, that described point in time exists, would be to waitIdle right before the first timestamp. And then it shouldn't matter if it's TOP or BOTTOM.

Reply:
I'm not sure I understand this concern:
> with BOTTOM to BOTTOM, you can be sure to not have any work from previous commands on your bill, but might have missed some of your work.

How could some of the render-pass work not be included given BOTTOM is the last stage in the pipeline and the spec for vkCmdWriteTimestamp says:
```The first synchronization scope includes all commands that occur earlier in submission order. The synchronization scope is limited to operations on the pipeline stage specified by pipelineStage.```

I'd think all work in the render-pass must be included in the timestamp's first sync scope here, as for something to _not_ be included it would have to execute _after_ the BOTTOM stage -- which is not possible? (That said, a possible concern is that _other_ work, after the render-pass, could be done _before_ the timestamp (so long as the timestamp's BOTTOM stage execute before the other works' BOTTOM stage). To avoid this, one could e.g. emit a BOTTOM -> TOP pipeline barrier directly after the barrier, however.)

Reply:
Ah, right, I think I see what you mean now :) The BOTTOM timestamp _before_ the render-pass is not guaranteed to be taken before the render-pass starts. What do you think about adding a pipe-barrier as a workaround for this?

In summary:
- Timestamp in BOTTOM (i.e. after all previous work)
- BOTTOM -> TOP barrier (guarantees timestamp is taken before the upcoming render-pass starts)
- Render-pass
- Timestamp in BOTTOM (i.e. after all previous work)
- .. repeat for each render-pass ..

Reply:
Yes, with such a barrier, you would make sure that the timestamp is taken after all previous work and before the work to be measured. And it shouldn't matter then, whether you start at BOTTOM or at TOP.
But of course, you would stall the GPU.
@SaschaWillems: after all this discussion here, it seems worthwhile to have a few words on this topic in the tutorial as well.

Reply:
... and maybe such barriers can be used to get more accurate timing results on the single stages as well?
I think, that's definitely worth to be tried.

Reply:
Thanks everyone. That's a valuable discussion :)

But I'd rather not add barriers in between.  I just took a look how tracy, a very popular CPU/GPU profiler, uses timestamp queries and it seems to be at least similar to what I'm doing. With the exception that tracy always only uses the BOTTOM stage bit. But no barriers.